File: src/main/java/scripting/quest/QuestScriptManager.java
package scripting.quest;

import client.Client;
import lombok.Getter;
import scripting.AbstractScriptManager;

import javax.script.Invocable;
import java.util.HashMap;
import java.util.Map;

public class QuestScriptManager extends AbstractScriptManager {

    private final @Getter Map<Client, QuestConversationManager> qms = new HashMap<>();
    private final Map<Client, Invocable> scripts = new HashMap<>();
    private static QuestScriptManager instance;

    public static QuestScriptManager getInstance() {
        if (instance == null) {
            instance = new QuestScriptManager();
        }
        return instance;
    }

    private QuestScriptManager() {

    }

    public void converse(Client c, int npc, int questId, boolean start) {
        converse(c, npc, questId, start, 1, 0);
    }

    public void converse(Client c, int mode, int selection) {
        QuestConversationManager qm = qms.get(c);
        converse(c, qm.getNpcId(), qm.getId(), qm.isStart(), mode, selection);
    }

    public boolean converse(Client c, int npc, int questId, boolean start, int mode, int selection) {
        //MapleQuest quest = MapleQuest.getInstance(questid);
        /*if (!c.getPlayer().getQuest(quest).getStatus().equals(MapleQuestStatus.Status.NOT_STARTED)) {
            dispose(c);
            return;
        }*/
        try {
            if (scripts.get(c) == null) {
                QuestConversationManager qm = new QuestConversationManager(c, npc, questId, start);
                if (qms.containsKey(c)) {
                    dispose(c);
                }
                if (c.canClickNPC()) {
                    qms.put(c, qm);
                    Invocable iv = getInvocable("quest/" + questId + ".js", c);
                    if (iv == null) {
                        System.out.println("Quest " + questId + " is uncoded.");
                        return false;
                    }
                    engine.put("cm", qm);
                    scripts.put(c, iv);
                    c.setLastNpcClick(System.currentTimeMillis());
                    iv.invokeFunction(start ? "start" : "end", mode, selection);
                } else {
                    c.getCharacter().enableActions();
                }
            } else {
                c.setLastNpcClick(System.currentTimeMillis());
                scripts.get(c).invokeFunction(start ? "start" : "end", mode, selection);
            }
        } catch (final Exception e) {
            System.err.println(e.getMessage());
            dispose(c);
            return false;
        }
        return true;
    }

    public void dispose(QuestConversationManager qm, Client c) {
        qms.remove(c);
        scripts.remove(c);
        resetContext("quest/" + qm.getId() + ".js", c);
    }

    public void dispose(Client c) {
        QuestConversationManager qm = qms.get(c);
        if (qm != null) {
            dispose(qm, c);
        }
    }
}

File: src/main/java/scripting/quest/QuestScriptManager.kt
package scripting.quest

import client.Client
import scripting.AbstractScriptManager
import javax.script.Invocable

object QuestScriptManager : AbstractScriptManager() {

    val qms: MutableMap<Client, QuestConversationManager> = HashMap()
    private val scripts: MutableMap<Client, Invocable> = HashMap()

    fun converse(c: Client, npc: Int, questId: Int, start: Boolean) {
        converse(c, npc, questId, start, 1, 0)
    }

    fun converse(c: Client, mode: Int, selection: Int) {
        val qm = qms[c] ?: return
        converse(c, qm.npcId, qm.id, qm.isStart, mode, selection)
    }

    fun converse(c: Client, npc: Int, questId: Int, start: Boolean, mode: Int, selection: Int): Boolean {
        //MapleQuest quest = MapleQuest.getInstance(questid);
        /*if (!c.getPlayer().getQuest(quest).getStatus().equals(MapleQuestStatus.Status.NOT_STARTED)) {
            dispose(c);
            return;
        }*/
        try {
            if (scripts[c] == null) {
                val qm = QuestConversationManager(c, npc, questId, start)
                if (qms.containsKey(c)) {
                    dispose(c)
                }
                if (c.canClickNPC()) {
                    qms[c] = qm
                    val iv = getInvocable("quest/$questId.js", c)

                    if (iv == null) {
                        println("Quest $questId is uncoded.")
                        return false
                    }

                    val unmutableEngine = engine ?: return false
                    unmutableEngine.put("cm", qm)
                    scripts[c] = iv

                    c.lastNpcClick = System.currentTimeMillis()
                    iv.invokeFunction(if (start) "start" else "end", mode, selection)
                } else {
                    c.character.enableActions()
                }
            } else {
                c.lastNpcClick = System.currentTimeMillis()
                scripts[c]!!.invokeFunction(if (start) "start" else "end", mode, selection)
            }
        } catch (e: Exception) {
            //System.err.println(e.getMessage());
            e.printStackTrace()
            dispose(c)
            return false
        }
        return true
    }

    fun dispose(qm: QuestConversationManager, c: Client) {
        qms.remove(c)
        scripts.remove(c)
        resetContext("quest/" + qm.id + ".js", c)
    }

    fun dispose(c: Client) {
        val qm = qms[c] ?: return
        dispose(qm, c)
    }
}